### PR TITLE
Pos validator uptime

### PR DESCRIPTION
**Pull request by arturrez**
## Why this should be merged
adds  PoS proof of uptime support for `removeValitor` 

## How this works
if uptime is not provided, it's calculated from evm getCurrentValdators api.
Signature aggregator is used to sign proof of uptime msg which is submitted to ValidatorManager smart contract
## How this was tested
```
./bin/avalanche node local start  --num-nodes 7 --etna-devnet local7
./bin/avalanche blockchain create pos7 --proof-of-stake  --force  --evm --reward-basis-points 10000000 // make sure that validator gets reward even after short staking period
./bin/avalanche blockchain deploy pos7 --cluster local7 --bootstrap-endpoints http://127.0.0.1:9650/,http://127.0.0.1:9652/,http://127.0.0.1:9654/,http://127.0.0.1:9656/,http://127.0.0.1:9658/  --convert-only
./bin/avalanche node local track local7 pos7
./bin/avalanche contract initValidatorManager pos7  --cluster local7
./bin/avalanche blockchain addValidator pos7  --cluster local7 --node-endpoint http://127.0.0.1:9660
sleep 120s
./bin/avalanche blockchain removeValidator pos7  --cluster local7 --node-endpoint http://127.0.0.1:9660
```

### CODE REVIEW

File: cmd/blockchaincmd/remove_validator.go [hunk 1]
```diff
@@ -199,11 +208,23 @@ func removeValidator(_ *cobra.Command, args []string) error {
 	return nil
 }
 
+func isBootstrapValidatorForNetwork(nodeID ids.NodeID, scNetwork models.NetworkData) bool {
+	filteredBootstrapValidators := utils.Filter(scNetwork.BootstrapValidators, func(b models.SubnetValidator) bool {
+		if id, err := ids.NodeIDFromString(b.NodeID); err == nil && id == nodeID {
```
[the remark this answers is not in the document]
Comment (arturrez) [reply cmd/blockchaincmd/remove_validator.go:213]: logic is to check if nodeID belong to the list of validators.
There is no point of raising an error as we would just ignore for incorrect nodeID as incorrect NodeID would not be in the list of bootstrap validator NodeIDs.

File: cmd/blockchaincmd/remove_validator.go [hunk 2]
```diff
@@ -261,13 +287,40 @@ func removeValidatorSOV(
 		aggregatorAllowPrivatePeers,
 		aggregatorLogLevel,
 		sc.PoS(),
-		force,
+		uptimeSec,
+		isBootstrapValidator || force,
 	)
-	if err != nil {
+	if err != nil && errors.Is(err, validatormanagerSDK.ErrValidatorIneligibleForRewards) {
+		ux.Logger.PrintToUser("Validator %s is not eligible for rewards", nodeID)
```
[the remark this answers is not in the document]
Comment (arturrez) [reply cmd/blockchaincmd/remove_validator.go:294]: improved msg

File: pkg/utils/net.go [hunk 3]
```diff
@@ -65,3 +67,26 @@ func IsValidIPPort(ipPortPair string) bool {
 	}
 	return true
 }
+
+// SplitRPCURI splits the RPC URI into `endpoint` and `chain`.
+// Reverse operation of `fmt.Sprintf("%s/ext/bc/%s", endpoint, chain)`.
+// returns the `uri` and `chain` as strings, or an error if the request URI is invalid.
+// SplitRPCURI splits the RPC URL into `endpoint` and `chain`.
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/net.go:74]: fixed

File: pkg/utils/net.go [hunk 4]
```diff
@@ -65,3 +67,26 @@ func IsValidIPPort(ipPortPair string) bool {
...
+// It matches the pattern `http://<endpoint>/ext/bc/<chain>/rpc` using a regex.
+func SplitRPCURI(requestURI string) (string, string, error) {
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/net.go:76]: agree, but suggest to keep it like that for now as it's only used to parse local rpc endpoints that are exposed by `avalanchego`

[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/net.go:76]: I think I should rename this func to reflect it's limited usage

File: pkg/utils/staking.go [hunk 5]
```diff
@@ -73,3 +78,41 @@ func GetNodeParams(nodeDir string) (
...
+	}
+	return 0, errors.New("nodeID not found in validator set: " + nodeID.String())
+}
+
+// GetL1ValidatorUptimeSeconds returns the uptime of the L1 validator
+func GetL1ValidatorUptimeSeconds(rpcURL string, nodeID ids.NodeID) (uint64, error) {
+	const uptimeDeductible = uint64(10) // seconds to make sure all L1 validators would agree on uptime
+	ctx, cancel := GetAPIContext()
+	defer cancel()
+	networkEndpoint, blockchainID, err := SplitRPCURI(rpcURL)
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/staking.go:104]: I renamed a function to reflect it's limited usage

File: pkg/utils/staking.go [hunk 6]
```diff
@@ -73,3 +78,41 @@ func GetNodeParams(nodeDir string) (
...
+	if err != nil {
+		return 0, err
+	}
+	evmCli := subnetEvmPlugin.NewClient(networkEndpoint, blockchainID)
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/staking.go:108]: agree.  I would be happy if `SplitAvalanchegoRPCURI` would also addressed there as these 2 func rely on exact format of RPC url and reverse each other functionality in terms of rpc URL

File: pkg/utils/staking.go [hunk 7]
```diff
@@ -73,3 +78,41 @@ func GetNodeParams(nodeDir string) (
...
+		if v.NodeID == nodeID {
+			return time.Unix(int64(v.EndTime), 0).Sub(startTime), nil
+		}
+	}
+	return 0, errors.New("nodeID not found in validator set: " + nodeID.String())
+}
+
+// GetL1ValidatorUptimeSeconds returns the uptime of the L1 validator
+func GetL1ValidatorUptimeSeconds(rpcURL string, nodeID ids.NodeID) (uint64, error) {
+	const uptimeDeductible = uint64(10) // seconds to make sure all L1 validators would agree on uptime
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/staking.go:101]: moved to constants

File: pkg/utils/staking.go [hunk 8]
```diff
@@ -73,3 +78,41 @@ func GetNodeParams(nodeDir string) (
...
+	if err != nil {
+		return 0, err
+	}
+	evmCli := subnetEvmPlugin.NewClient(networkEndpoint, blockchainID)
+	validators, err := evmCli.GetCurrentValidators(ctx, []ids.NodeID{nodeID})
+	if err != nil {
+		return 0, err
+	}
+	if len(validators) > 0 {
+		return validators[0].UptimeSeconds - uptimeDeductible, nil
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/staking.go:114]: it's not possible as minimun staking period is 100 seconds and uptime can't be below this

File: pkg/utils/net_test.go [hunk 9]
```diff
@@ -29,3 +29,71 @@ func TestIsValidIPPort(t *testing.T) {
...
+			expectedChain:    "",
+			expectError:      true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			endpoint, chain, err := SplitRPCURI(tt.requestURI)
+
+			if tt.expectError {
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/net_test.go:82]: fixed

File: pkg/utils/staking.go [hunk 10]
```diff
@@ -4,15 +4,20 @@ package utils
...
+	"time"
 
 	"github.com/ava-labs/avalanche-cli/pkg/constants"
 	"github.com/ava-labs/avalanchego/ids"
 	"github.com/ava-labs/avalanchego/staking"
 	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+	"github.com/ava-labs/avalanchego/vms/platformvm"
 	"github.com/ava-labs/avalanchego/vms/platformvm/signer"
+
+	subnetEvmPlugin "github.com/ava-labs/subnet-evm/plugin/evm"
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/staking.go:20]: agree. fixed

File: pkg/utils/staking.go [hunk 11]
```diff
@@ -73,3 +78,41 @@ func GetNodeParams(nodeDir string) (
 	}
 	return nodeID, blsPub, blsPoP, nil
 }
+
+func GetValidationTime(networkEndpoint string, nodeID ids.NodeID, subnetID ids.ID, startTime time.Time) (time.Duration, error) {
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/staking.go:82]: done

File: pkg/validatormanager/validatormanager.go [hunk 12]
```diff
@@ -90,6 +91,7 @@ func AddRewardCalculatorToAllocations(
 	allocs core.GenesisAlloc,
 	rewardBasisPoints uint64,
 ) {
+	ux.Logger.PrintToUser("Adding reward calculator contract to allocations %d", rewardBasisPoints)
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/validatormanager/validatormanager.go:94]: sorry leaked debug output. removed

File: pkg/utils/staking.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (felipemadero) [reply pkg/utils/staking.go]: can you add a check anyway

**Review by felipemadero — APPROVED** [review]

**Pull request merged**